utils/ensemble_tools.py

```diff
@@ -0,0 +1,159 @@
+        scan_points[parameter] = []
+        for i in range(n_dirs):
+            scan_points[parameter].append(param_value)
+            list_all.append("{0}_{1}".format(parameter, i))
+            param_value += (
+                parameter_dict[parameter][1] - parameter_dict[parameter][0]
+            ) / float(n_dirs - 1)
+
+    # Create the combinations of these parameters.
+    combinations = list(itertools.combinations(list_all, len(parameter_dict.keys())))
+    all_configs = []
+    for comb in combinations:
+        record_comb = True
+        for i in range(len(comb) - 1):
+            search_term = comb[i][: len(comb[i]) - comb[i][::-1].find("_")]
+            for j in range(i + 1, len(comb)):
+                if comb[j].startswith(search_term):
+                    record_comb = False
+                    break
+        if record_comb:
+            all_configs.append(comb)
+
+    # If destructive, let's delete the whole tree if it already exists
+    if destructive and os.path.isdir(sweep_path):
+        shutil.rmtree(sweep_path)
+    # Check whether the copy directory exists
+    copy_files = os.path.isdir(copy_dir)
+    # Now loop over each combination, create a directory, copy the
+    # configs and conditions, and edit the configuation to contain
+    # the correct parameters.
+    for comb in all_configs:
+        dirName = "-".join(comb)
+        dir_path = "{0}/SWEEP/{1}".format(sweep_path, dirName)
+        make_directory(dir_path, destructive)
+        if copy_files:
+            copy_dir_contents(dir_path, copy_dir)
+        # Create parameter map for conditions encoding
+        parameters = {}
+        for parameter in parameter_dict.keys():
+            param_index = [item for item in comb if item.startswith(parameter)][
+                0
+            ].split("_")[-1]
+            parameters[parameter] = scan_points[parameter][int(param_index)]
+        if edit_file:
+            encode_conditions_file(Path(dir_path) / edit_file, parameters)
+    return 0
+
+
+def copy_dir_contents(dir_path, copy_dir):
+    """Copy the contents of copy_dir -> dir_path"""
+
+    for f in os.listdir(copy_dir):
+        shutil.copy(copy_dir + f, dir_path)
+
+
+def make_directory(directory_name, destructive):
+    """Make the directory tree at directory_name."""
+
+    if os.path.isdir(directory_name):
+        if destructive:
+            shutil.rmtree(directory_name)
+        else:
+            print("Path already exists and " "not in destructive mode! Returning")
+            return 0
+    os.makedirs(directory_name)
+    return 1
+
+
+def encode_conditions_file(in_file_name, param_dict):
+    """Encode a configuration file with a dict of input values"""
+
+    for param_name in param_dict.keys():
+        edit_parameter(in_file_name, param_name, param_dict[param_name])
+
+
+def edit_parameter(in_file_name, param, val):
+    """Edit a single parameter in the configuration
+    file to the desired value"""
```

Comment:
The approach here for changing parameter values is a very brittle. For example I think if the `param` string appears anywhere else in the file other than the parameter definition we'll still end up substituting this with something, which will probably break things. I would say either better to use a combination of parsing XML and querying for relevant node and using a regular expression or similar to parse the parameter definition as in https://github.com/UCL/neso-calibration/blob/52e63174f6347d501f7f62771107949be4e299d4/src/nesopy/model.py#L121 or using a templated conditions file would be better from a robustness and flexibility perspective.

Reply:
I was a little worried about this method too - I realised that, for example, the hw.xml has an 's' parameter that would certainly lead to undesirable consequences if we were to try and template it.

I've implemented a version for pattern matching the XML parsed parameters using a slightly adjusted version of the method above. I'm using lxml because it preserves comments in the XML file by default.

Reply:
Using `lxml` is fine though [it looks it is may be possible to use `xml` module in standard library and preserve comments](https://stackoverflow.com/a/66668127). We will need to add a `requirements.txt` to the root of the repository with `lxml` listed if using to ensure it is installed when the plug-in is.

Reply:
I've used the method linked, so have gone back to the regular ```xml ``` package.